Qt: widget for entering amounts, supports both BTC and sats

kristapsk opened the pull request:
![image](https://user-images.githubusercontent.com/4500994/75630663-7833d780-5bf5-11ea-87de-ce3bc19893e4.png)

It has dropdown to switch between entering BTC or sats. From caller's perspective tries to behave like QLineEdit and always returns sats.

Currently used only in single join / send form. Should be added to other places, like Tumbler wizard too, but that takes extra work as QWizard does not know how to handle them by default and also needs more testing.

Changed code that creates controls in single send from loops to adding each control seperately. It's more readable and also gives more flexibility with using different controls in future (for example, mixdepth chooser could be drop down with mixdepth number and available balance in brackets).

Commented out unimplemented SchStaticPage with line comments, because """ approach sometimes messes up Vim's syntax highlightning.

kristapsk commented:
Rebased

kristapsk commented:
Rebased again. Planning to work on the remaining stuff mentioned above, but some concept ACKs / NACKs would not hurt.

chris-belcher commented:
Concept ACK

AdamISZ commented:
First, sorry that this one's been a bit neglected!
Ran a few simple tests, this is what I observed:

* the UI looks good, simple and intuitive toggle.

* did coinjoins with btc and sat amounts and worked as intended

* counter-intuitive behaviour: you *can* enter non-integer sats values, and the pre-existing auto-correction to btc occurs. the same thing happens if you enter a large (>21M) integer value for BTC. So basically the settings work but are not enforced. IIRC you can use something like `QIntValidator` or similar in Qt to enforce the right kind of input, there is some usage of that already, somewhere in joinmarket-qt.py or qtsupport.py.

Once the above is cleared up I'll check it again. I do think it's a good addition to the app though, thanks.

kristapsk commented:
@AdamISZ , thanks for testing, will check the bug you mentioned (I really think it's a bug, not just "counter-intuitive behaviour"). Also still haven't managed to get tumbler wizards working with this, want to do that too.

If this gets merged before, should be useful for #487 too.

kristapsk commented:
Rebased and addressed comments by @AdamISZ. Currently works in a single send and settings tabs. Not yet in tumbler wizard. I tried to get that working some weeks ago, didn't finish, will get back to it at some point, but I think this is already useful, even without that.

AdamISZ commented:
Test result:
I got a stack trace when clicking out of the "address" entry field to the next (entered a valid regtest address). It appears not to be related to the PR, but rather to the BIP21 thing?

```
(jmvenv) waxwing@here~/testjminstall/joinmarket-clientserver/scripts$ python joinmarket-qt.py --datadir=.
User data location: .
2020-05-18 16:22:25,680 [DEBUG]  rpc: getnewaddress []
2020-05-18 16:22:25,712 [DEBUG]  rpc: sendrawtransaction ['0000000000000000000000000000000000000000']
2020-05-18 16:22:25,714 [DEBUG]  error pushing = -22 TX decode failed
2020-05-18 16:22:38,789 [DEBUG]  rpc: listaddressgroupings []
2020-05-18 16:22:38,830 [DEBUG]  Fast sync in progress. Got this many used addresses: 4
2020-05-18 16:22:40,119 [DEBUG]  rpc: listunspent [0]
2020-05-18 16:22:40,229 [DEBUG]  bitcoind sync_unspent took 0.11186861991882324sec
2020-05-18 16:22:40,281 [INFO]  Starting transaction monitor in walletservice
Traceback (most recent call last):
  File "joinmarket-qt.py", line 486, in <lambda>
    lambda: checkAddress(self, self.addressInput.text()))
  File "joinmarket-qt.py", line 119, in checkAddress
    parent.widgets[0][1].setText(addr)
AttributeError: 'SpendTab' object has no attribute 'widgets'
```
At a first glance I don't know where this came from; it does indeed appear that `QWidget` does not have a `widgets` member variable? https://doc.qt.io/qt-5/qwidget.html

AdamISZ commented on scripts/joinmarket-qt.py:
```diff
@@ -506,12 +477,44 @@ def initUI(self):
 
         donateLayout = self.getDonateLayout()
         innerTopLayout.addLayout(donateLayout, 0, 0, 1, 2)
-        self.widgets = getSettingsWidgets()
```
Looks like this is the cause of the stack trace, so I guess a fairly trivial fix (refer to the object itself, `SpendTab.widgets` no longer exists ... I see you already made this edit elsewhere, so I guess just an oversight.)

AdamISZ commented on scripts/joinmarket-qt.py:
```diff
@@ -527,7 +530,7 @@ def initUI(self):
         buttons.addWidget(self.startButton)
         buttons.addWidget(self.abortButton)
         self.abortButton.clicked.connect(self.abortTransactions)
-        innerTopLayout.addLayout(buttons, len(self.widgets) + 1, 0, 1, 2)
```
It's a shame to lose flexibility by hardcoding 5. But a really minor point so OK.

AdamISZ commented on scripts/qtsupport.py:
```diff
@@ -99,7 +100,7 @@
     "console_log_level": "one of INFO, DEBUG, WARN, ERROR; INFO is least noisy;\n" +
     "consider switching to DEBUG in case of problems.",
-    "absurd_fee_per_kb": "maximum satoshis/kilobyte you are willing to pay,\n" +
+    "absurd_fee_per_kb": "maximum amount per kilobyte you are willing to pay,\n" +
```
Why this change? I don't get it.

kristapsk replied:
Because you can now enter either BTC or sats there.

AdamISZ replied:
I see. Hadn't noticed that it was added there. Indeed it's a nice addition, unfortunate that `tx_fees` being a more complex setting cannot be done the same way. Still, good.

AdamISZ commented on scripts/qtsupport.py:
```diff
@@ -507,20 +508,109 @@ def filter(self, p, columns):
-    def __init__(self, parent):
-        super(SchStaticPage, self).__init__(parent)
-        self.setTitle("Manually create a schedule entry")
+# TODO implement this option
```
Why change this in this PR?

kristapsk replied:
Mentioned in initial comment when opening PR. It messes up syntax highlighting in vim sometimes, makes no difference otherwise. Can be removed if you insist.

AdamISZ replied:
To be clear: you're saying that using `"""` style comment sections is a problem for formatting sometimes? You think that should always be avoided outside of doc strings, then? Not unreasonable, if so, just I wasn't aware.

kristapsk replied:
Yes, it looks so. Can't say with 100% certainty, vim syntax highlighting acts strange sometimes. Also `"""` is meant for doc strings, `#` is a proper way to comment out code blocks, I think.

kristapsk commented:
Addressed stack trace mentioned by @AdamISZ, my mistake at rebasing. Had this in mind while doing #567, but forgot.

AdamISZ commented:
Testing: https://github.com/JoinMarket-Org/joinmarket-clientserver/commit/2c54542bd4eeac566adb7ba117eb607da8005663

I notice something that is probably a mistake: I can enter more than 8 decimal places for BTC amounts (specifically, 15 decimal places). From trying a coinjoin, I can confirm it simply rounds to the nearest 8 decimal places, as you'd expect, but this is not the correct behaviour imo: users should be constrained to 8 decimal places for an exact amount (and indeed all other wallets do this).

AFAIR this can be set in a `QValidator`.

Apart from this point, everything else so far looks correct, including rejection of invalid input (negative or non-numeric input). Also the number of significant figures allowed for satoshi amounts looks correct, and non-integers are correctly rejected.

I'll test a little more but the above point seems like it needs to be fixed.

kristapsk commented:
Good catch, `BitcoinAmountBTCValidator` should be improved, it also currently allows entering multiple decimal points, looking into it.

AdamISZ commented on scripts/qtsupport.py:
```diff
@@ -507,20 +508,115 @@ def filter(self, p, columns):
+class BitcoinAmountBTCValidator(QDoubleValidator):
+
+    def __init__(self):
+        super().__init__(0.00000000, 20999999.9769, 8)
```
Given my report here: https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/530#issuecomment-630755454 this looks like something weird is going on, you set the accuracy correctly ...

kristapsk replied:
I override `validate()` method, that's why it didn't work. :)

kristapsk commented:
Fixed

AdamISZ commented:
Fix confirmed to above with https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/530/commits/55f7f5f166056e17f079abc93f2884dfc915a9cb

Testing autoconvert, I notice a behaviour which I don't think was intended: 34 sats converts to 3.4E-7 BTC.

kristapsk commented:
Fixed E notation issue, the same fix as in #573.

AdamISZ commented:
For future reference, it's better to add commits one at a time and then squash at the end. That way I don't need to spend time finding the new code that fixes each bug.

AdamISZ commented:
I'm guessing https://github.com/JoinMarket-Org/joinmarket-clientserver/commit/6a659ac8d758397cd575b73a8900ca070fe3dd45 added this? https://github.com/JoinMarket-Org/joinmarket-clientserver/commit/6a659ac8d758397cd575b73a8900ca070fe3dd45#diff-3e361749da5986d8e0337a42d1eb755bR598

kristapsk commented:
> I'm guessing [6a659ac](https://github.com/JoinMarket-Org/joinmarket-clientserver/commit/6a659ac8d758397cd575b73a8900ca070fe3dd45) added this? [6a659ac#diff-3e361749da5986d8e0337a42d1eb755bR598](https://github.com/JoinMarket-Org/joinmarket-clientserver/commit/6a659ac8d758397cd575b73a8900ca070fe3dd45#diff-3e361749da5986d8e0337a42d1eb755bR598)

Yes

AdamISZ commented:
OK tACK, I will merge this.

AdamISZ commented:
Sorry, before I merge: you have tabs on empty lines at line 645 and 650, should be removed (don't know if other places, didn't check every line.

kristapsk commented:
Fixed. These two were the only places, checked with `git diff HEAD~`.

AdamISZ commented:
I'm still seeing them in https://github.com/JoinMarket-Org/joinmarket-clientserver/commit/2645c1b44fc70fa1b6afbd4d952c71bbd39874f8 ?

kristapsk commented:
I forgot to `git add` before commit. :)
